<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>chevreulProcess â€¢ chevreulProcess</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="chevreulProcess">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">chevreulProcess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.23</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/chevreulProcess.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cobriniklab/chevreulProcess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>chevreulProcess</h1>
                        <h4 data-toc-skip class="author">Kevin
Stachelek</h4>
            <address class="author_afil">
      University of Southern
California<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevin.stachelek@gmail.com" class="email">kevin.stachelek@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Bhavana
Bhat</h4>
            <address class="author_afil">
      University of Southern
California<br><a class="author_email" href="mailto:#"></a><a href="mailto:bbhat@usc.edu" class="email">bbhat@usc.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">26 December 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/cobriniklab/chevreulProcess/blob/master/vignettes/chevreulProcess.Rmd" class="external-link"><code>vignettes/chevreulProcess.Rmd</code></a></small>
      <div class="d-none name"><code>chevreulProcess.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="basics">Basics<a class="anchor" aria-label="anchor" href="#basics"></a>
</h2>
<div class="section level3">
<h3 id="install-chevreulprocess">Install <code>chevreulProcess</code><a class="anchor" aria-label="anchor" href="#install-chevreulprocess"></a>
</h3>
<p><code>R</code> is an open-source statistical environment which can be
easily modified to enhance its functionality via packages. <em><a href="https://bioconductor.org/packages/3.20/chevreulProcess" class="external-link">chevreulProcess</a></em>
is a <code>R</code> package available via the <a href="http://bioconductor.org" class="external-link">Bioconductor</a> repository for packages.
<code>R</code> can be installed on any operating system from <a href="https://cran.r-project.org/" class="external-link">CRAN</a> after which you can install
<em><a href="https://bioconductor.org/packages/3.20/chevreulProcess" class="external-link">chevreulProcess</a></em>
by using the following commands in your <code>R</code> session:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"chevreulProcess"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="required-knowledge">Required knowledge<a class="anchor" aria-label="anchor" href="#required-knowledge"></a>
</h3>
<p>The <em><a href="https://bioconductor.org/packages/3.20/chevreulProcess" class="external-link">chevreulProcess</a></em>
package is designed for single-cell RNA sequencing data. The functions
included within this package are derived from other packages that have
implemented the infrastructure needed for RNA-seq data processing and
analysis. Packages that have been instrumental in the development of
<em><a href="https://bioconductor.org/packages/3.20/chevreulProcess" class="external-link">chevreulProcess</a></em>
include, <code>Biocpkg("SummarizedExperiment")</code> and
<code>Biocpkg("scater")</code>.</p>
</div>
<div class="section level3">
<h3 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a>
</h3>
<p><code>R</code> and <code>Bioconductor</code> have a steep learning
curve so it is critical to learn where to ask for help. The <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support site</a>
is the main resource for getting help: remember to use the
<code>chevreulProcess</code> tag and check <a href="https://support.bioconductor.org/tag/chevreulProcess/" class="external-link">the older
posts</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="quick-start-to-using-chevreulprocess">Quick start to using <code>chevreulProcess</code><a class="anchor" aria-label="anchor" href="#quick-start-to-using-chevreulprocess"></a>
</h2>
<p>The <code>chevreulProcess</code> package contains functions to
preprocess, cluster, visualize, and perform other analyses on scRNA-seq
data. It also contains a shiny app for easy visualization and analysis
of scRNA data.</p>
<p><code>chvereul</code> uses SingelCellExperiment (SCE) object type
(from <em><a href="https://bioconductor.org/packages/3.20/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>)
to store expression and other metadata from single-cell experiments.</p>
<p>This package features functions capable of:</p>
<ul>
<li>Performing Clustering at a range of resolutions and Dimensional
reduction of Raw Sequencing Data.</li>
<li>Visualizing scRNA data using different plotting functions.</li>
<li>Integration of multiple datasets for consistent analyses.</li>
<li>Cell cycle state regression and labeling.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/cobriniklab/chevreulProcess" class="external-link">"chevreulProcess"</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in get(paste0(generic, ".", class), envir = get_method_env()) : </span></span>
<span><span class="co">#&gt;   object 'type_sum.accel' not found</span></span>
<span></span>
<span><span class="co"># Load the data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cobriniklab/chevreuldata" class="external-link">chevreuldata</a></span><span class="op">)</span></span>
<span><span class="va">chevreul_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/chevreuldata/man/human_gene_transcript_sce.html" class="external-link">human_gene_transcript_sce</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">chevreul_sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 56267 794 </span></span>
<span><span class="co">#&gt; metadata(4): merge.info pca.info experiment markers</span></span>
<span><span class="co">#&gt; assays(3): reconstructed counts logcounts</span></span>
<span><span class="co">#&gt; rownames(56267): 5-8S-rRNA 5S-rRNA ... ZZEF1 ZZZ3</span></span>
<span><span class="co">#&gt; rowData names(1): rotation</span></span>
<span><span class="co">#&gt; colnames(794): hs20151130-SC1-26 hs20151130-SC1-28 ...</span></span>
<span><span class="co">#&gt;   20200312-DS-dissected-81 20200312-DS-dissected-83</span></span>
<span><span class="co">#&gt; colData names(33): batch Sequencing_Run ... gene_snn_res.0.8</span></span>
<span><span class="co">#&gt;   gene_snn_res.1</span></span>
<span><span class="co">#&gt; reducedDimNames(3): corrected TSNE UMAP</span></span>
<span><span class="co">#&gt; mainExpName: integrated</span></span>
<span><span class="co">#&gt; altExpNames(2): gene transcript</span></span></code></pre></div>
<p><code>R</code> session information.</p>
<pre><code><span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] chevreuldata_0.99.25        ExperimentHub_2.14.0       </span></span>
<span><span class="co">#&gt;  [3] AnnotationHub_3.14.0        BiocFileCache_2.14.0       </span></span>
<span><span class="co">#&gt;  [5] dbplyr_2.5.0                chevreulProcess_0.99.23    </span></span>
<span><span class="co">#&gt;  [7] scater_1.34.0               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [9] scuttle_1.16.0              SingleCellExperiment_1.28.1</span></span>
<span><span class="co">#&gt; [11] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt; [13] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        </span></span>
<span><span class="co">#&gt; [15] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt; [17] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      </span></span>
<span><span class="co">#&gt; [19] matrixStats_1.4.1           BiocStyle_2.34.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] jsonlite_1.8.9            shape_1.4.6.1            </span></span>
<span><span class="co">#&gt;   [3] magrittr_2.0.3            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;   [5] GenomicFeatures_1.58.0    rmarkdown_2.29           </span></span>
<span><span class="co">#&gt;   [7] GlobalOptions_0.1.2       fs_1.6.5                 </span></span>
<span><span class="co">#&gt;   [9] BiocIO_1.16.0             zlibbioc_1.52.0          </span></span>
<span><span class="co">#&gt;  [11] ragg_1.3.3                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [13] memoise_2.0.1             Rsamtools_2.22.0         </span></span>
<span><span class="co">#&gt;  [15] DelayedMatrixStats_1.28.0 RCurl_1.98-1.16          </span></span>
<span><span class="co">#&gt;  [17] htmltools_0.5.8.1         S4Arrays_1.6.0           </span></span>
<span><span class="co">#&gt;  [19] curl_6.0.1                BiocNeighbors_2.0.1      </span></span>
<span><span class="co">#&gt;  [21] SparseArray_1.6.0         sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [23] bslib_0.8.0               desc_1.4.3               </span></span>
<span><span class="co">#&gt;  [25] cachem_1.1.0              ResidualMatrix_1.16.0    </span></span>
<span><span class="co">#&gt;  [27] GenomicAlignments_1.42.0  igraph_2.1.2             </span></span>
<span><span class="co">#&gt;  [29] mime_0.12                 lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;  [31] pkgconfig_2.0.3           rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [33] Matrix_1.7-1              R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [35] fastmap_1.2.0             GenomeInfoDbData_1.2.13  </span></span>
<span><span class="co">#&gt;  [37] digest_0.6.37             colorspace_2.1-1         </span></span>
<span><span class="co">#&gt;  [39] AnnotationDbi_1.68.0      dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [41] irlba_2.3.5.1             textshaping_0.4.1        </span></span>
<span><span class="co">#&gt;  [43] RSQLite_2.3.9             beachmat_2.22.0          </span></span>
<span><span class="co">#&gt;  [45] filelock_1.0.3            httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [47] abind_1.4-8               compiler_4.4.2           </span></span>
<span><span class="co">#&gt;  [49] bit64_4.5.2               withr_3.0.2              </span></span>
<span><span class="co">#&gt;  [51] BiocParallel_1.40.0       viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [53] DBI_1.2.3                 rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [55] DelayedArray_0.32.0       rjson_0.2.23             </span></span>
<span><span class="co">#&gt;  [57] bluster_1.16.0            tools_4.4.2              </span></span>
<span><span class="co">#&gt;  [59] vipor_0.4.7               beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [61] glue_1.8.0                restfulr_0.0.15          </span></span>
<span><span class="co">#&gt;  [63] batchelor_1.22.0          grid_4.4.2               </span></span>
<span><span class="co">#&gt;  [65] cluster_2.1.6             generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [67] megadepth_1.16.0          gtable_0.3.6             </span></span>
<span><span class="co">#&gt;  [69] tzdb_0.4.0                ensembldb_2.30.0         </span></span>
<span><span class="co">#&gt;  [71] hms_1.1.3                 metapod_1.14.0           </span></span>
<span><span class="co">#&gt;  [73] BiocSingular_1.22.0       ScaledMatrix_1.14.0      </span></span>
<span><span class="co">#&gt;  [75] XVector_0.46.0            BiocVersion_3.20.0       </span></span>
<span><span class="co">#&gt;  [77] stringr_1.5.1             ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt;  [79] pillar_1.10.0             limma_3.62.1             </span></span>
<span><span class="co">#&gt;  [81] circlize_0.4.16           dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [83] lattice_0.22-6            rtracklayer_1.66.0       </span></span>
<span><span class="co">#&gt;  [85] bit_4.5.0.1               tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [87] locfit_1.5-9.10           Biostrings_2.74.1        </span></span>
<span><span class="co">#&gt;  [89] knitr_1.49                gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [91] bookdown_0.41             ProtGenerics_1.38.0      </span></span>
<span><span class="co">#&gt;  [93] edgeR_4.4.1               cmdfun_1.0.2             </span></span>
<span><span class="co">#&gt;  [95] xfun_0.49                 statmod_1.5.0            </span></span>
<span><span class="co">#&gt;  [97] stringi_1.8.4             UCSC.utils_1.2.0         </span></span>
<span><span class="co">#&gt;  [99] EnsDb.Hsapiens.v86_2.99.0 lazyeval_0.2.2           </span></span>
<span><span class="co">#&gt; [101] yaml_2.3.10               evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [103] codetools_0.2-20          tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [105] BiocManager_1.30.25       cli_3.6.3                </span></span>
<span><span class="co">#&gt; [107] systemfonts_1.1.0         munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [109] jquerylib_0.1.4           Rcpp_1.0.13-1            </span></span>
<span><span class="co">#&gt; [111] png_0.1-8                 XML_3.99-0.17            </span></span>
<span><span class="co">#&gt; [113] parallel_4.4.2            pkgdown_2.1.1            </span></span>
<span><span class="co">#&gt; [115] readr_2.1.5               blob_1.2.4               </span></span>
<span><span class="co">#&gt; [117] scran_1.34.0              AnnotationFilter_1.30.0  </span></span>
<span><span class="co">#&gt; [119] sparseMatrixStats_1.18.0  bitops_1.0-9             </span></span>
<span><span class="co">#&gt; [121] viridisLite_0.4.2         scales_1.3.0             </span></span>
<span><span class="co">#&gt; [123] purrr_1.0.2               crayon_1.5.3             </span></span>
<span><span class="co">#&gt; [125] rlang_1.1.4               KEGGREST_1.46.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kevin Stachelek, Bhavana Bhat.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
